src/recovery.rs

```diff
@@ -532,35 +577,43 @@ impl Recovery {
             }
         }
 
-        if !lost_pkt.is_empty() {
-            self.on_packets_lost(lost_pkt, epoch, now);
+        if let Some(pkt) = largest_lost_pkt {
+            self.on_packets_lost(lost_bytes, &pkt, epoch, now);
         }
-    }
-
-    fn on_packet_acked(
-        &mut self, pkt_num: u64, epoch: packet::Epoch, now: Instant,
-    ) -> bool {
-        // Check if packet is newly acked.
-        if let Some(mut p) = self.sent[epoch].remove(&pkt_num) {
-            self.acked[epoch].append(&mut p.frames);
 
-            if p.in_flight {
-                self.on_packet_acked_cc(epoch, &p, now);
+        self.drain_packets(epoch);
+    }
 
-                self.delivery_rate.on_ack_received(p, now);
+    fn drain_packets(&mut self, epoch: packet::Epoch) {
+        let mut lowest_non_expired_pkt_index = self.sent[epoch].len();
+
+        // In order to avoid removing elements from the middle of the list
+        // (which would require copying other elements to compact the list),
+        // we only remove a contguous range of elements from the start of the
+        // list.
+        //
+        // This means that acked or lost elements coming after this will not
+        // be removed at this point, but their removal is delayed for a later
+        // time, once the gaps have been filled.
+
+        // First, find the first element that is neither acked not lost.
```

Comment:
nit: grammar of `neither acked not lost`

Reply:
Fixed, thanks.